### PR TITLE
*: add, clean up e2e tests

### PR DESCRIPTION
**Pull request by gyuho**

### CODE REVIEW

**Comment by gyuho** [comment]
1. added test: `auth enable`, `defrag`, `endpoint status`, `member list`
2. cleaned up by:
 - separate each command to each go file
 - separate `curl` command tests in v2 into a separate file

TODO: `alarm`, `auth disable (not implemented)`, `snapshot`, `user`, `role`, `member (other commands)`

/cc @heyitsanthony @xiang90 Thanks.

File: e2e/ctl_v2_curl_test.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
+// Copyright 2016 CoreOS, Inc.
```
Comment (heyitsanthony) [inline e2e/ctl_v2_curl_test.go]: Should this file be prefixed with `ctl`? It's not using etcdctl.

File: e2e/ctl_v2_curl_test.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+package e2e
+
+import (
+	"math/rand"
+	"testing"
+
+	"github.com/coreos/etcd/pkg/testutil"
+)
+
+func TestBasicOpsNoTLS(t *testing.T)        { testBasicOpsPutGet(t, &configNoTLS) }
```
Comment (heyitsanthony) [inline e2e/ctl_v2_curl_test.go]: Should these be `TestV2Curl...`?

File: e2e/ctl_v3_defrag_test.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+	"testing"
+)
+
+func TestCtlV3Defrag(t *testing.T) { testCtl(t, defragTest) }
+
+func defragTest(cx ctlCtx) {
+	var kvs = []kv{{"key", "val1"}, {"key", "val2"}, {"key", "val3"}}
+	for i := range kvs {
+		if err := ctlV3Put(cx, kvs[i].key, kvs[i].val, ""); err != nil {
+			cx.t.Fatalf("compactTest #%d: ctlV3Put error (%v)", i, err)
```
Comment (heyitsanthony) [inline e2e/ctl_v3_defrag_test.go]: `cx.t.Fatal(err)`

File: e2e/ctl_v3_defrag_test.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+		}
+	}
+
```
Comment (heyitsanthony) [inline e2e/ctl_v3_defrag_test.go]: should there be a compact?

File: e2e/ctl_v3_defrag_test.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+	if err := ctlV3Defrag(cx); err != nil {
+		cx.t.Fatalf("defragTest ctlV3Defrag error (%v)", err)
+	}
+}
+
+func ctlV3Defrag(cx ctlCtx) error {
+	cmdArgs := append(cx.PrefixArgs(), "defrag")
+	lines := make([]string, cx.epc.cfg.clusterSize)
+	for i, ep := range cx.epc.endpoints() {
+		lines[i] = fmt.Sprintf("Finished defragmenting etcd member[%s]", ep)
```
Comment (heyitsanthony) [inline e2e/ctl_v3_defrag_test.go]: just `Finished defragmenting etcd member`? Defrag shouldn't be serialized like it is now-- output ordering could change.

File: e2e/ctl_v3_endpoint_test.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import "testing"
+
+func TestCtlV3EndpointHealthQuorum(t *testing.T) { testCtl(t, endpointHealthTest, withQuorum()) }
+func TestCtlV3EndpointStatusQuorum(t *testing.T) { testCtl(t, endpointStatusTest, withQuorum()) }
```
Comment (heyitsanthony) [inline e2e/ctl_v3_endpoint_test.go]: `TestCtlV3EndpointStatus` -- there's no quorum in this command

File: e2e/etcd_test.go [hunk 7]
```diff
@@ -464,3 +370,18 @@ func (epc *etcdProcessCluster) proxies() []*etcdProcess {
 func (epc *etcdProcessCluster) backends() []*etcdProcess {
 	return epc.procs[:epc.cfg.clusterSize]
 }
+
+func (epc *etcdProcessCluster) endpoints() []string {
+	eps := make([]string, epc.cfg.clusterSize)
+	for i, pc := range epc.backends() {
+		ep := pc.cfg.acurl
+		if strings.HasPrefix(ep, "http://") {
```
Comment (heyitsanthony) [inline e2e/etcd_test.go]: I don't think the `HasPrefix` checks are necessary.

File: e2e/ctl_v3_endpoint_test.go (hunk 6 above)
Comment (gyuho) [inline e2e/ctl_v3_endpoint_test.go]: Oh will fix. thanks.

File: e2e/ctl_v3_defrag_test.go (hunk 5 above)
Comment (gyuho) [inline e2e/ctl_v3_defrag_test.go]: You are right. will fix.

**Comment by gyuho** [comment]
@heyitsanthony PTAL.

1. renamed the `curl` test file with `v2_curl_test`
2. change `curl` test names with `TestV2Curl`
3. added compaction operation before defrag command
4. changed `defrag` command test to just check slice of strings without endpoints in it because it's not in order
5. delete `HasPrefix` parts and just use `url.Parse.Host` to get the endpoints without scheme

Thanks.

File: e2e/ctl_v3_test.go [hunk 8]
```diff
@@ -119,12 +72,12 @@ func withInteractive() ctlOption {
 	return func(cx *ctlCtx) { cx.interactive = true }
 }
 
-func setupCtlV3Test(t *testing.T, cfg etcdProcessClusterConfig, quorum bool) *etcdProcessCluster {
+func setupCtlV3Test(t *testing.T, cfg *etcdProcessClusterConfig, quorum bool) *etcdProcessCluster {
```
Comment (heyitsanthony) [inline e2e/ctl_v3_test.go]: why the change to `*etcdProcessClusterConfig`?

File: e2e/ctl_v3_defrag_test.go [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+	}
+
+	if err := ctlV3Defrag(cx); err != nil {
+		cx.t.Fatalf("defragTest ctlV3Defrag error (%v)", err)
+	}
+}
+
+func ctlV3Defrag(cx ctlCtx) error {
+	cmdArgs := append(cx.PrefixArgs(), "defrag")
+	lines := make([]string, cx.epc.cfg.clusterSize)
```
Comment (gyuho) [inline e2e/ctl_v3_defrag_test.go]: @heyitsanthony I changed it to pass pointer because otherwise this returns 3 even when it's not running with quorum.

Comment (heyitsanthony) [inline e2e/ctl_v3_defrag_test.go]: So the quorum flag isn't being respected? I don't think that is a good way to fix the problem.

Comment (gyuho) [inline e2e/ctl_v3_defrag_test.go]: It is respected, but we just pass around the preconfigured config object and to not manipulate that shared config, we pass them by value in setupCtlV3Test function. I will try to clean up arg types so that we don't do this hacky stuff. Thanks!

Comment (heyitsanthony) [inline e2e/ctl_v3_defrag_test.go]: I think the fix is to move the code in `setupCtlV3Test` into `testCtl` so ret.cfg is set when the quorum flag is checked.

Comment (gyuho) [inline e2e/ctl_v3_defrag_test.go]: Ok good idea. Let me try that. Thanks!

**Comment by gyuho** [comment]
@heyitsanthony Just removed `setupCtlV3Test`. PTAL. Thanks!

**Comment by heyitsanthony** [comment]
lgtm

**Comment by gyuho** [comment]
thanks. merging in greens.
